staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -265,18 +460,196 @@ func (q *delayingType) waitingLoop() {
+	existing, exists := knownEntries[entry.data]
+	if exists {
+		if q.logger != nil {
+			q.log("waitingLoop: existing found: " + existing.String())
+		}
+	}
+
+	// the returnCh can be used to make a call synchronous or inspect 'waiting' queue state.
+	if entry.returnCh != nil {
+		defer close(entry.returnCh)
+
+		// return if the client is only interested in the existence check.
+		if entry.features.check_existence {
+			if exists {
+				ret = 1
+			}
+			if q.logger != nil {
+				q.log(fmt.Sprintf("waitingLoop: return exists result %t for %v", exists, entry.data))
+			}
+			entry.returnCh <- ret
+			return
+		}
+
+		// return if the client is only interested in waiting_length_check
+		if entry.features.check_length {
+			ret = len(knownEntries)
+			if q.logger != nil {
+				q.log("waitingLoop: waiting queue length: " + strconv.Itoa(ret))
+			}
+			entry.returnCh <- ret
+			return
+		}
+	}
+
+	// Support the removal of items from the waiting queue, by request.
+	if entry.features.forget {
+		if exists {
+			// remove the item from the 'waiting' queue
+			if q.logger != nil {
+				q.log(fmt.Sprintf("waitingLoop: waiting entry forgotten: %v", entry.data))
+			}
+
+			head := w.Peek().(*waitFor)
+			heap.Remove(w, existing.index)
+			delete(knownEntries, entry.data)
+			// if we forgot the head item recalculate the nextReadyAt timer.
+			if existing == head {
+				q.calculateNextReadyAt(w)
+			}
 		}
+		return
+	}
 
+	// Perform dedupe between the 'waiting' queue against items in the 'active' queue.
+	// Dedup is disabled when PermitActiveAndWaiting is set in the options.
+	if !entry.options.PermitActiveAndWaiting && q.IsQueued(entry.data) {
+		if q.logger != nil {
+			q.log(fmt.Sprintf("waitingLoop: already active queued drops: %v", entry.data))
+		}
 		return
 	}
 
-	heap.Push(q, entry)
-	knownEntries[entry.data] = entry
+	// drop if the incoming item if the caller wishes to keep the existing item if it exists.
+	if exists && entry.options.WhenWaiting == TakeExisting {
+		if q.logger != nil {
+			q.log(fmt.Sprintf("waitingLoop: TakeExisting and already waiting drops: %v", entry.data))
+		}
+		return
+	}
+
+	// delayed add (to 'waiting' queue)
+	if entry.readyAt.After(q.clock.Now()) {
+		// inform the metrics of the retry (moved from AddAfter())
+		q.metrics.retry()
+
+		// lookup the existing head readyAt time so that we can check if we need to recalculate it after the insert
+		var headReady time.Time
+		if len(knownEntries) > 0 {
+			headReady = w.Peek().(*waitFor).readyAt
+		}
+		// insert the entry into the 'waiting' queue with appropriate insertion strategy defined by options.Waiting
+		// re-calculate nextReadyAt when we insert the first item into the queue or replace the item at the head.
+		if insert(w, knownEntries, entry) {
+			if q.logger != nil {
+				q.log("waitingLoop: inserted into waiting queue: " + entry.String())
+			}
+			if len(knownEntries) == 1 || headReady.After(entry.readyAt) {
```

Comment:
Please see my comments https://github.com/kubernetes/kubernetes/pull/112328#issuecomment-1428011201
I changed it because there was a race which could effect the existing code but was less likely because the Add, AddAfter methods executed more quickly.

The main point is we need to not allow the Synchronous add to complete until the item has been received AND new nextReadyAt calculated.  Note, we "could" try and move the closing of the return channel out to the waitingLoop but this will be complicated to understand.

The original implementation recalculates the nextReadyAt all of the time, and 50% ish of the time its not needed because the event didn't change the head item.  Of course, every time the head item matures and becomes ready then we're going to have a new head (or no head).

Reply:
> The main point is we need to not allow the Synchronous add to complete until the item has been received AND new nextReadyAt calculated. Note, we "could" try and move the closing of the return channel out to the waitingLoop but this will be complicated to understand.

As shown in https://github.com/kubernetes/kubernetes/pull/112328#issuecomment-1443853545, we might not need to do that. But we could still preserve this behaviour with option 3 if desired. It seems to me, that this should not be that complicated. We would just close the channel if it is present after calculating the nextReadyAtTimer. We could write a comment explaining why we do this to make it clearer. 

As a positive consequence, it would make the processArrivingEntry function more transparent and readable, when we move both calculateNextReadyAt and returnCh closing out.

> The original implementation recalculates the nextReadyAt all of the time, and 50% ish of the time its not needed because the event didn't change the head item. Of course, every time the head item matures and becomes ready then we're going to have a new head (or no head).

ok, we can skip the original option 1. WDYT about the options 2. and 3. ?